### ISSUE TITLE
[ComponentTableViewCell] Wrong height

### ISSUE THREAD

**Issue by zdnk** [alexdrone/Render#68]
Hello,

since TableNode does not support UIRefreshControl for pull to refresh I tried using regular UITableView with ComponentTableViewCell but the height of the cells was not calculated properly :(

**Comment by zdnk** [comment]
What about this? It is still an issue

**Comment by alexdrone** [comment]
I can't reproduce the issue - can you attach some code pls? :)

**Comment by zdnk** [comment]
Sure, here is my component:
```swift
import Foundation
import Render
import UIKit

fileprivate let baseSpacing: CGFloat = 12
fileprivate let corners: CGFloat = 2

class GenericApprovalsListItemComponentView: ComponentView<GenericApprovalsListItemState> {
    
    var isLast: Bool = false
    
    override func render() -> NodeType {
        let container = Node<UIView>(reuseIdentifier: "list-item-container") { [unowned self] view, layout, size in
            layout.width = size.width
            layout.paddingTop = 10
            layout.paddingLeft = 10
            layout.paddingRight = 10
            layout.paddingBottom = self.isLast ? 10 : 0
            
            view.backgroundColor = .clear
        }
        
        let innerContainer = Node<UIView>(reuseIdentifier: "list-item-innerContainer") { view, layout, size in
            layout.percent.width = 100%
            layout.padding = baseSpacing
            layout.flexDirection = .column
            layout.flex()
            
            view.backgroundColor = .white
            view.cornerRadius = corners
            view.layer.zPosition = -1
        }
        
        let titlesColumn = Node<UIView>(reuseIdentifier: "list-item-titleColumn") { view, layout, size in
            layout.flexDirection = .column
            layout.justifyContent = .center
            layout.flex()
            }.add(children: [
                titleNode(),
                subtitleNode(),
            ])
        
        let topContainer = Node<UIView>(reuseIdentifier: "list-item-topContainer") { view, layout, size in
            layout.flexDirection = .row
            layout.alignItems = .center
            layout.flex()
            }.add(children: [
                iconNode(),
                titlesColumn,
            ])
        
        var blocks: [NodeType] = []
        
        if let currency = state.currency, let amount = state.amount, !currency.isEmpty, !amount.isEmpty {
            blocks.append(amountNode())
        }
        
        if let desc = state.itemDescription, !desc.isEmpty {
            blocks.append(descriptionNode())
        }
        
        if state.attachments.count > 0 {
            blocks.append(attachmentsListNode())
        }
        
        if let note = state.note, !note.isEmpty {
            blocks.append(noteNode())
        }
        
        if blocks.count > 0 {
            let spacer = Node<UIView> { _, layout, _ in
                layout.height = baseSpacing
            }
            
            blocks.insert(spacer, at: 0)
        }
        
        innerContainer.add(children: [topContainer] + blocks)
        
        var outerBlocks: [NodeType] = []
        
        if state.escalation != .none {
            outerBlocks.append(contentsOf: escalationNodes())
        }
        
        outerBlocks.append(innerContainer)
        
        if let completion = state.completion {
            outerBlocks.append(contentsOf: completionNodes(completion))
        }
        
        return container.add(children: outerBlocks)
    }
    
}

fileprivate extension GenericApprovalsListItemComponentView {
    
    func iconNode() -> NodeType {
        return Node<UIImageView>(reuseIdentifier: "list-item-icon") { [unowned self] view, layout, size in
            layout.marginRight = baseSpacing
            layout.height = 60
            layout.width = layout.height
            
            view.clipsToBounds = true
            view.image = UIImage(named: self.state.iconName)
        }
    }
    
    func titleNode() -> NodeType {
        return Node<UILabel>(reuseIdentifier: "list-item-title") { [unowned self] label, layout, size in
            layout.marginBottom = 3
            
            label.text = self.state.title
            label.font = Theme.ListItem.Font.title
            label.textColor = Theme.ListItem.Color.title
            label.numberOfLines = 0
            label.lineBreakMode = .byWordWrapping
        }
    }
    
    func subtitleNode() -> NodeType {
        return Node<UILabel>(reuseIdentifier: "list-item-subtitle") { [unowned self] label, layout, size in
            label.text = self.state.subtitle
            label.font = Theme.ListItem.Font.subtitle
            label.textColor = Theme.ListItem.Color.subtitle
            label.numberOfLines = 0
            label.lineBreakMode = .byWordWrapping
        }
    }
    
    func attachmentsListNode() -> NodeType {
        let container = Node<UIView>(reuseIdentifier: "list-item-attachments-container") { view, layout, size in
            layout.flexDirection = .column
            layout.flex()
        }
        
        let nodes = state.attachments.map { (attachment) -> NodeType in
            return ComponentNode(
                GenericApprovalsListItemAttachmentComponentView(),
                in: self,
                reuseIdentifier: "list-item-attachment",
                key: "list-item-\(self.state.id)-attachment-\(attachment.name.hash)",
                state: attachment
            )
        }
        
        return container.add(children: nodes)
    }
    
    func amountNode() -> NodeType {
        return Node<UIView>(reuseIdentifier: "list-item-amountRow") { view, layout, size in
            layout.flexDirection = .row
            layout.justifyContent = .flexEnd
            layout.marginBottom = baseSpacing
            layout.flex()
        }.add(children: [
            Node<UILabel>(reuseIdentifier: "list-item-amountRow-currency") { [unowned self] label, layout, size in
                layout.marginRight = 4
                
                label.text = self.state.currency
                label.font = Theme.ListItem.Font.currency
                label.textColor = Theme.ListItem.Color.currency
            },
            Node<UILabel>(reuseIdentifier: "list-item-amountRow-amount") { [unowned self] label, layout, size in
                label.text = self.state.amount
                label.font = Theme.ListItem.Font.amount
                label.textColor = Theme.ListItem.Color.amount
            },
        ])
    }
    
    func descriptionNode() -> NodeType {
        return Node<UILabel>(reuseIdentifier: "list-item-itemDescription") { [unowned self] label, layout, size in
            label.numberOfLines = 2
            label.lineBreakMode = .byWordWrapping
            label.text = self.state.itemDescription
            label.font = Theme.ListItem.Font.itemDescription
            label.textColor = Theme.ListItem.Color.itemDescription
        }
    }
    
    func noteNode() -> NodeType {
        return Node<UILabel>(reuseIdentifier: "list-item-note") { [unowned self] label, layout, size in
            layout.marginTop = 4
            
            label.numberOfLines = 2
            label.lineBreakMode = .byWordWrapping
            label.text = self.state.note
            label.font = Theme.ListItem.Font.note
            label.textColor = Theme.ListItem.Color.note
        }
    }
    
    func escalationNodes() -> [NodeType] {
        let stripe = Node<UIView>(reuseIdentifier: "list-item-escalation-stripe") { [unowned self] view, layout, size in
            layout.percent.width = 100%
            layout.height = 20
            layout.flexDirection = .row
            layout.alignItems = .center
            layout.paddingLeft = baseSpacing
            
            view.backgroundColor = self.state.escalation == .escalated ? Theme.ListItem.Color.escalated : Theme.ListItem.Color.reminded
        }.add(children: [
            Node<UIImageView>(reuseIdentifier: "list-item-escalation-icon") { [unowned self] view, layout, size in
                layout.height = 28
                layout.width = layout.height
                layout.marginRight = 4
                
                view.cornerRadius = layout.height / 2
                view.backgroundColor = self.state.escalation == .escalated ? Theme.ListItem.Color.escalated : Theme.ListItem.Color.reminded
                view.image = UIImage(named: self.state.escalation == .escalated ? "Escalated" : "Reminder")
                view.contentMode = .center
            },
            Node<UILabel>(reuseIdentifier: "list-item-escalation-label") { [unowned self] label, layout, size in
                label.text = self.state.escalation == .escalated ? "ESCALATED" : "REMINDED"
                label.font = Theme.ListItem.Font.escalation
                label.textColor = Theme.ListItem.Color.escalation
            },
        ])
        
        return [
            stripe,
            Node<UIView>(reuseIdentifier: "list-item-escalation-bottom") { [unowned self] view, layout, size in
                layout.percent.width = 100%
                layout.height = 4
                layout.marginBottom = -4
                
                view.layer.zPosition = -2
                view.backgroundColor = self.state.escalation == .escalated ? Theme.ListItem.Color.escalated : Theme.ListItem.Color.reminded
            },
        ]
    }
    
    func completionNodes(_ completion: GenericApprovalsListItemState.Completion) -> [NodeType] {
        let top = Node<UIView>(reuseIdentifier: "list-item-completion-top") { view, layout, size in
            layout.percent.width = 100%
            layout.height = 8
            layout.marginTop = -4
            layout.marginBottom = -4
            
            view.layer.zPosition = -2
            view.backgroundColor = Theme.ListItem.Color.completionBg
        }
        
        let container = Node<UIView>(reuseIdentifier: "list-item-completion-container") { view, layout, size in
            layout.percent.width = 100%
            layout.height = 46
            layout.flexDirection = .row
            layout.alignItems = .center
            layout.paddingLeft = baseSpacing
            layout.paddingRight = baseSpacing
            layout.flex()
            
            view.cornerRadius = corners
            view.backgroundColor = Theme.ListItem.Color.completionBg
        }
        
        let approvalIcon = Node<UIImageView>(reuseIdentifier: "list-item-completion-icon") { view, layout, size in
            layout.height = 29
            layout.width = 29
            layout.marginRight = baseSpacing
            
            view.image = UIImage(named: "ApprovalApprovedIcon")
        }
        
        let approvalLabel = Node<UILabel>(reuseIdentifier: "list-item-completion-label") { label, layout, size in
            layout.flexGrow = 1
            layout.flexShrink = 1
            
            label.font = Theme.ListItem.Font.completion
            label.textColor = Theme.ListItem.Color.completion
            label.text = "Approved on " + completion.completedOn
        }
        
        container.add(children: [approvalIcon, approvalLabel])
        
        return [top, container]
    }
    
}

class GenericApprovalsListItemAttachmentComponentView: ComponentView<GenericApprovalsListItemState.Attachment> {
    
    override func render() -> NodeType {
        let container = Node<UIView>(reuseIdentifier: "list-item-attachment") { view, layout, size in
            layout.flexDirection = .row
            layout.percent.width = 100%
            layout.flexGrow = 1
            layout.flexShrink = 1
            layout.alignItems = .center
            layout.flex()
        }
        
        return container.add(children: [
            iconNode(),
            nameNode(),
        ])
    }
    
}

fileprivate extension GenericApprovalsListItemAttachmentComponentView {
    
    func iconNode() -> NodeType {
        return Node<UIImageView>(reuseIdentifier: "list-item-attachment-icon") { view, layout, size in
            layout.height = 26
            layout.width = 22
            layout.marginRight = 8
            
            let image = UIImage(named: "MidasAttachmentIcon")
            view.image = image
            view.clipsToBounds = true
        }
    }
    
    func nameNode() -> NodeType {
        return Node<UILabel>(reuseIdentifier: "list-item-attachment-name") { [unowned self] label, layout, size in
            layout.flexGrow = 1
            layout.flexShrink = 1
            
            label.text = self.state.name
            label.font = Theme.ListItem.Attachment.Font.name
            label.textColor = Theme.ListItem.Attachment.Color.name
            label.numberOfLines = 0
            label.lineBreakMode = .byWordWrapping
        }
    }
    
}
```
State:
```swift
import Foundation
import Render

struct GenericApprovalsListItemState: StateType {
    
    struct Attachment: StateType {
        
        let name: String

    }
    
    struct Completion {
        
        let decision: ApprovalDecision
        let comment: String?
        let completedOn: String
        
    }
    
    let id: String
    let title: String
    let subtitle: String
    let iconName: String
    let currency: String?
    let amount: String?
    let itemDescription: String?
    let note: String?
    let escalation: ListItemEscalationType
    let attachments: [Attachment]
    let isDownloaded: Bool
    let completion: Completion?
    
}

extension GenericApprovalsListItemState.Attachment {
    
    init() {
        name = ""
    }
    
}

extension GenericApprovalsListItemState.Attachment: Equatable {
    
    static func ==(lhs: GenericApprovalsListItemState.Attachment, rhs: GenericApprovalsListItemState.Attachment) -> Bool {
        return lhs.name == rhs.name
    }
    
}

extension GenericApprovalsListItemState.Completion: Equatable {
    
    static func ==(lhs: GenericApprovalsListItemState.Completion, rhs: GenericApprovalsListItemState.Completion) -> Bool {
        return (
            lhs.comment == rhs.comment &&
            lhs.decision == rhs.decision &&
            lhs.completedOn == rhs.completedOn
        )
    }
    
}

extension GenericApprovalsListItemState {
    
    init() {
        id = ""
        title = ""
        subtitle = ""
        iconName = ""
        currency = nil
        amount = nil
        itemDescription = nil
        note = nil
        escalation = .none
        attachments = []
        isDownloaded = false
        completion = nil
    }
    
}

extension GenericApprovalsListItemState: Equatable {
    
    static func ==(lhs: GenericApprovalsListItemState, rhs: GenericApprovalsListItemState) -> Bool {
        let a: Bool = (
            lhs.id == rhs.id &&
            lhs.title == rhs.title &&
            lhs.subtitle == rhs.subtitle &&
            lhs.iconName == rhs.iconName &&
            lhs.currency == rhs.currency &&
            lhs.amount == rhs.amount &&
            lhs.itemDescription == rhs.itemDescription &&
            lhs.note == rhs.note &&
            lhs.escalation == rhs.escalation &&
            lhs.isDownloaded == rhs.isDownloaded
        )
        
        return a && lhs.attachments == rhs.attachments
    }
    
}
```

Then I am just using regular `UITableView` with regular `.dataSource` pattern to register `ComponentTableViewCell` and dequeue reusable cells. Can you test this please?

**Comment by alexdrone** [comment]
Have you set 

```
    tableView.rowHeight = UITableViewAutomaticDimension
    tableView.estimatedRowHeight = 64  ... (-1 on iOS 11)
```
in your tableView?

**Comment by zdnk** [comment]
Yes I did

**Comment by zdnk** [comment]
Ok, it now just works when I tried it again .. am I stupid or what :/

**Comment by alexdrone** [comment]
Can I close :P ?

**Comment by zdnk** [comment]
Yes, thanks. However, it would be nice to solve the UIResfreshControl with TableNode instead of using plain UITableView.

**Comment by zdnk** [comment]
So, it is still not working after all. I was using modified version of ComponentTableViewCell. I added these:
```swift
    public override func sizeThatFits(_ size: CGSize) -> CGSize {
        return component.sizeThatFits(size)
    }
    
    public override func systemLayoutSizeFitting(_ targetSize: CGSize) -> CGSize {
        return (component as! UIView).systemLayoutSizeFitting(targetSize)
    }
```

without them I get 64pt cell height for every cell

**Comment by alexdrone** [comment]
This is now fixed in Render 4.9 :)
